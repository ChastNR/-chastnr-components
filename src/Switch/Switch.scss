.sw {
  position: relative;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  border-radius: 40px;
  cursor: pointer;
  display: inline-block;
  height: 24px;
  outline: none;
  overflow: hidden;
  text-align: left;
  user-select: none;
  vertical-align: middle;
  width: 48px;

  &.sw__checked {
    &::before {
      background-color: #ffdd2d;
    }

    & > span {
      &::after {
        transform: translate3d(24px, 0, 0);
      }

      span:last-child {
        left: 0;
      }
    }
  }

  &::before {
    background-color: #ecf1f7;
    border-radius: inherit;
    content: "";
    height: 100%;
    left: 0;
    pointer-events: none;
    position: absolute;
    top: 0;
    transition: background-color 0.25s ease, opacity 0.25s ease;
    width: 100%;
  }

  & > span {
    span:first-child {
      left: 0;
    }

    span:last-child {
      right: 0;
    }
  }

  .sw__box {
    border-radius: inherit;
    display: block;
    height: inherit;
    outline: none;
    width: inherit;

    .sw__box__left {
      position: absolute;
      top: 0;
      bottom: 0;
      box-sizing: border-box;
      opacity: 0;
      color: #333;
      transition-duration: 0.25s;
      transition-property: transform, opacity;
      transform: scale(0.66);
      will-change: transform, opacity;
      margin-left: 2px;
      display: flex;
      width: 50%;
      justify-content: center;
      align-items: center;
    }

    .sw__box__right {
      position: absolute;
      top: 0;
      bottom: 0;
      box-sizing: border-box;
      opacity: 0;
      color: #333;
      transition-duration: 0.25s;
      transition-property: transform, opacity;
      transform: scale(0.66);
      will-change: transform, opacity;
      border-style: solid;
      border-radius: 50%;
      border-top-color: transparent;
      margin: 4px;
      width: 16px;
      height: 16px;
      border-width: 2px;
    }

    &::before {
      border-radius: inherit;
      content: "";
      height: 100%;
      left: 0;
      position: absolute;
      top: 0;
      transition: background-color 0.25s ease;
      width: 100%;
    }

    &::after {
      background-color: #fff;
      border-radius: 100%;
      content: "";
      height: 16px;
      margin: 4px;
      position: absolute;
      transition: transform 0.25s ease;
      width: 16px;
    }
  }
}
