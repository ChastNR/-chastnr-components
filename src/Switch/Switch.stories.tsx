import { useState } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { ComponentMeta } from '@storybook/react';
import { Switch } from './Switch';

const componentMeta: ComponentMeta<typeof Switch> = {
  title: 'Switch',
  component: Switch,
  parameters: {
    layout: 'fullscreen',
  },
};

export const DefaultSwitch = () => {
  const [isChecked, setCheched] = useState(false);

  const handleChange = () => setCheched((s) => !s);

  return <Switch checked={isChecked} name="testName" onChange={handleChange} />;
};

export default componentMeta;
