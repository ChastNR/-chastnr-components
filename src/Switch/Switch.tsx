import classNames from "clsx";

import "./Switch.scss";

interface SwitchProps {
  name?: string;
  checked?: boolean;
  onChange?: React.MouseEventHandler<HTMLSpanElement>;
}

const Switch: React.FC<SwitchProps> = ({ name, checked, onChange }) => {
  const switchClassNames = classNames("sw", {
    sw__checked: checked,
  });

  return (
    <span className={switchClassNames} onClick={onChange}>
      <span className="sw__box" tabIndex={-1}>
        <input checked={checked} hidden name={name} type="checkbox" />
        <span className="sw__box__left" />
        <span className="sw__box__right" />
      </span>
    </span>
  );
};

export default Switch;
