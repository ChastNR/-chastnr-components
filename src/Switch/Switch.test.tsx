import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Switch } from './Switch';

describe('<Switch/>', () => {
  it('should match snapshot', () => {
    const switchComponent = shallow(<Switch />);
    expect(toJson(switchComponent)).toMatchSnapshot();
  });

  it('props should be equal', () => {
    const name = 'testName';
    const checked = true;
    const onChange = jest.fn();

    const switchComponent = mount(<Switch checked={checked} onChange={onChange} name={name} />);

    const props = switchComponent.props();

    expect(props.name).toEqual(name);
    expect(props.checked).toEqual(checked);
  });
});
