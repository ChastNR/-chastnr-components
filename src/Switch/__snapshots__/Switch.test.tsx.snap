// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch/> should match snapshot 1`] = `
<span
  className="sw"
>
  <span
    className="sw__box"
    tabIndex={-1}
  >
    <input
      hidden={true}
      type="checkbox"
    />
    <span
      className="sw__box__left"
    />
    <span
      className="sw__box__right"
    />
  </span>
</span>
`;
